### PR TITLE
Use context to timeout kubelet query via httpClient and add timeout for the RestClient

### PR DESCRIPTION
**Pull request by kevinwangcn**
# Intent
To prevent the http query to kubelet from being stuck there for some reason

# Background
When the `Kubelet` service on a node is unhealthy or busy, the http query sent from kubeturbo may stuck there for a very long time, even set the `Timeout` on the `httpClient`, it still doesn't return for some reason. 

# Implementation
After doing some research about [Go Context](https://go.dev/blog/context) usage and referring to a [online guide](https://blog.cloudflare.com/the-complete-guide-to-golang-net-http-timeouts/), add a context with timeout to limit the time for a single kubelet query. 

# Test Done

## Case 1:  Use an online [http service](http://httpbin.org/delay/1) to test timeout context
**1. Sample code**
```
package main

import (
	"context"
	"fmt"
	"net/http"
	"time"
)

func main() {

	client := &http.Client{}

	//send 10 request on batch 0
	for i := 1; i <= 10; i++ {
		go sendrequst(client, i, 8, 0)
	}
	//send 10 request on batch 1
	for i := 1; i <= 10; i++ {
		go sendrequst(client, i, 5, 1)
	}

	println("wait in main ......")
	time.Sleep(60 * time.Second)

}

func sendrequst(c *http.Client, delay, timeout, batch int) {
	if delay > 10 {
		fmt.Println("You can't delay more than 10s with httpbin service!")
		return
	}

	req, _ := http.NewRequest(http.MethodGet, fmt.Sprintf("http://httpbin.org/delay/%v", delay), nil)
	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout)*time.Second)
	defer cancel()

	req = req.WithContext(ctx)
	resp, err := c.Do(req)

	if err != nil {
		fmt.Printf("batch[%v] task[%v] get error:%v\n", batch, delay, err) // cancel caught
	} else {
		fmt.Printf("batch[%v] task[%v] return code:%v\n", batch, delay, resp.StatusCode)
	}
}
```
**2. Run the test code and get result**
```
[root@IBM-PF3888SM go1proj]# go run httpclient.go 
wait in main ......
batch[0] task[1] return code:200
batch[1] task[1] return code:200
batch[0] task[2] return code:200
batch[1] task[2] return code:200
batch[0] task[3] return code:200
batch[1] task[3] return code:200
batch[1] task[4] return code:200
batch[0] task[4] return code:200
batch[1] task[6] get error:Get "http://httpbin.org/delay/6": context deadline exceeded
batch[1] task[5] get error:Get "http://httpbin.org/delay/5": context deadline exceeded
batch[1] task[10] get error:Get "http://httpbin.org/delay/10": context deadline exceeded
batch[1] task[9] get error:Get "http://httpbin.org/delay/9": context deadline exceeded
batch[1] task[7] get error:Get "http://httpbin.org/delay/7": context deadline exceeded
batch[1] task[8] get error:Get "http://httpbin.org/delay/8": context deadline exceeded
batch[0] task[5] return code:200
batch[0] task[6] return code:200
batch[0] task[7] return code:200
batch[0] task[8] get error:Get "http://httpbin.org/delay/8": context deadline exceeded
batch[0] task[9] get error:Get "http://httpbin.org/delay/9": context deadline exceeded
batch[0] task[10] get error:Get "http://httpbin.org/delay/10": context deadline exceeded
```

**3. Explanation**
Limit the timeout value to 8 in the batch 0, sending 10 requests,  getting 3 timeout task,  cause there are 3 requests whose time consumption is equal or great than 8 seconds

Limit the timeout value to 5 in the batch 1, sending 10 requests,  getting 5 timeout task,  cause there are 6 requests whose time consumption is equal or great than 5 seconds

### CODE REVIEW

File: pkg/kubeclient/kubelet_client.go [hunk 1]
```diff
@@ -178,8 +183,9 @@ func (client *KubeletClient) callKubeletEndpoint(ip, path string) ([]byte, error
 func (client *KubeletClient) callAPIServerProxyEndpoint(nodeName, path string) ([]byte, error) {
 	var statusCode int
 	fullPath := fmt.Sprintf("%s%s%s%s", "/api/v1/nodes/", nodeName, "/proxy", path)
-	body, err := client.kubeClient.CoreV1().RESTClient().Get().AbsPath(fullPath).Do(context.TODO()).StatusCode(&statusCode).Raw()
+	body, err := client.kubeClient.CoreV1().RESTClient().Get().Timeout(defaultConnTimeOut).AbsPath(fullPath).Do(context.TODO()).StatusCode(&statusCode).Raw()
```
Comment (pankajbatra1) [inline pkg/kubeclient/kubelet_client.go:186]: Please include discovery and monitoring results from lab targets.

File: pkg/kubeclient/kubelet_client.go [hunk 2]
```diff
@@ -155,9 +155,14 @@ func (client *KubeletClient) callKubeletEndpoint(ip, path string) ([]byte, error
...
 	}
 
+	ctx, cancel := context.WithTimeout(context.Background(), defaultConnTimeOut)
+	defer cancel()
+	req = req.WithContext(ctx)
+
 	httpClient := client.client
 	response, err := httpClient.Do(req)
 	if err != nil {
+		glog.V(3).Infof("Failed to query kubelet via httpClient for the node %v on the path %v as the error:%v", ip, path, err)
```
Comment (pankajbatra1) [inline pkg/kubeclient/kubelet_client.go:165]: Can you simulate some of these error conditions in lab or integration tests?

File: pkg/kubeclient/kubelet_client.go [hunk 3]
```diff
@@ -155,9 +155,14 @@ func (client *KubeletClient) callKubeletEndpoint(ip, path string) ([]byte, error
 		return nil, err
 	}
 
+	ctx, cancel := context.WithTimeout(context.Background(), defaultConnTimeOut)
```
Comment (chlam4) [inline pkg/kubeclient/kubelet_client.go:158]: Nit - As this technically isn't a connection timeout, let's not overload it?  We still can set it to the same 20 seconds, but let's use a different name?

**Review by chlam4 — APPROVED** [review]
Nice research/work @KevinWang0204!

**Comment by enlinxu** [comment]
Thanks for doing the research @KevinWang0204 . Feel free to share the learning in certain formats (quick presentation or wiki).

**Pull request merged**